File: packages/component-registry/src/seed.ts
import type { ComponentInventoryItem } from './types';

export const CATALOG_VERSION = '2026-04-25';

export const initialComponentInventory: ComponentInventoryItem[] = [
  {
    name: 'DateRangePicker',
    module: 'advanced-inputs',
    tag: 'DateRangePicker',
    pack: 'advanced-inputs',
    props: [
      { name: 'modelValue', type: 'string | [Date, Date] | null', required: false, description: 'Current selected date range.' },
      { name: 'placeholder', type: 'string', required: false, defaultValue: 'Start date - End date', description: 'Input placeholder text.' },
      { name: 'clearable', type: 'boolean', required: false, defaultValue: 'true', description: 'Show clear button.' },
      { name: 'format', type: 'string', required: false, defaultValue: 'YYYY-MM-DD', description: 'Date format passed to formatter.' },
    ],
    slots: [
      { name: 'header', description: 'Replace the calendar header area.' },
      { name: 'footer', description: 'Append custom action buttons.' },
    ],
    events: [
      { name: 'update:modelValue', payload: '{ start?: string; end?: string }', description: 'Emits when range changes.' },
      { name: 'apply', payload: '{ start: string; end: string }', description: 'Emits when user applies a range.' },
    ],
    examples: [
      {
        label: 'Report period chooser',
        pug: "DateRangePicker(v-model='reportRange' :clearable='true' format='YYYY-MM-DD')",
        description: 'Basic form usage in form groups.',
      },
    ],
    restrictions: [
      { type: 'runtime', message: 'Should not be used inside SSR-only rendering paths.' },
      { type: 'ux', message: 'Pair with clear reset controls when used in filter forms.' },
    ],
    enabled: true,
    version: '1.2.0',
  },
  {
    name: 'AsyncMultiSelect',
    module: 'advanced-inputs',
    tag: 'AsyncMultiSelect',
    pack: 'advanced-inputs',
    props: [
      { name: 'modelValue', type: 'string[]', required: false, description: 'Selected IDs.' },
      { name: 'search', type: '(query: string) => Promise<Option[]>', required: true, description: 'Async remote query function.' },
      { name: 'debounceMs', type: 'number', required: false, defaultValue: '300', description: 'Delay between remote requests.' },
      { name: 'labelKey', type: 'string', required: false, defaultValue: 'label', description: 'Option label field.' },
      { name: 'valueKey', type: 'string', required: false, defaultValue: 'value', description: 'Option value field.' },
    ],
    slots: [
      { name: 'option', description: 'Custom rendering for each suggestion item.' },
      { name: 'selected', description: 'Custom rendering for selected tokens.' },
    ],
    events: [
      { name: 'search', payload: 'string', description: 'Fires whenever query changes.' },
      { name: 'update:modelValue', payload: 'string[]', description: 'Selection changes.' },
    ],
    examples: [
      {
        label: 'Team member picker',
        pug: "AsyncMultiSelect(v-model='selectedUsers' :search='fetchUsers' label-key='fullName' value-key='id')",
        description: 'Load options as user types.',
      },
    ],
    restrictions: [
      { type: 'runtime', message: 'Requires a function prop that resolves within 15s.' },
      { type: 'security', message: 'Search input is sanitized through backend API proxy.' },
    ],
    enabled: true,
    version: '1.2.0',
  },
  {
    name: 'InputMask',
    module: 'advanced-inputs',
    tag: 'InputMask',
    pack: 'advanced-inputs',
    props: [
      { name: 'modelValue', type: 'string', required: false, description: 'Masked string value.' },
      { name: 'mask', type: 'string', required: true, description: 'Mask pattern, e.g. (999) 999-9999' },
      { name: 'placeholder', type: 'string', required: false, defaultValue: 'Enter value', description: 'Placeholder while empty.' },
      { name: 'unmask', type: 'boolean', required: false, defaultValue: 'false', description: 'Return unmasked value on input.' },
    ],
    slots: [{ name: 'addon', description: 'Custom input addon slot.' }],
    events: [
      { name: 'complete', payload: 'string', description: 'All mask characters completed.' },
      { name: 'update:modelValue', payload: 'string', description: 'Output value updates.' },
    ],
    examples: [
      {
        label: 'Phone field',
        pug: "InputMask(v-model='phone' mask='(999) 999-9999' placeholder='(201) 555-0123')",
        description: 'Use for strict format fields.' ,
      },
    ],
    restrictions: [
      { type: 'ux', message: 'Use with short fields and helper text for expected format.' },
      { type: 'security', message: 'Do not bind secrets or tokens to this component.' },
    ],
    enabled: true,
    version: '1.2.0',
  },
  {
    name: 'pm-table',
    module: 'packages/component-registry',
    tag: 'pm-table',
    pack: 'files',
    props: [
      { name: 'modelValue', type: 'Array<Record<string, any>>', required: false, description: 'Rows being displayed.' },
      { name: 'totalRows', type: 'number', required: false, description: 'Total matching rows from server.' },
      { name: 'sortBy', type: 'string[]', required: false, description: 'Sort columns order.' },
      { name: 'sortDesc', type: 'boolean[]', required: false, description: 'Sort directions.' },
      { name: 'currentPage', type: 'number', required: false, defaultValue: '1', description: 'Current page index.' },
      { name: 'perPage', type: 'number', required: false, defaultValue: '10', description: 'Page size.' },
      { name: 'filter', type: 'Record<string, any>', required: false, description: 'Server-side filter payload.' },
      { name: 'busy', type: 'boolean', required: false, defaultValue: 'false', description: 'Loading state.' },
    ],
    slots: [
      { name: 'cell', description: 'Custom cell rendering.' },
      { name: 'empty', description: 'Shown when dataset is empty.' },
    ],
    events: [
      { name: 'sort-change', payload: '{ sortBy: string[], sortDesc: boolean[] }', description: 'User changed sorting criteria.' },
      { name: 'filter-change', payload: 'Record<string, any>', description: 'User changed filter values.' },
      { name: 'page-change', payload: 'number', description: 'Page number changed.' },
      { name: 'update:items', payload: 'Array<Record<string, any>>', description: 'Updated data slice received from server.' },
    ],
    examples: [
      {
        label: 'Server paginated list',
        pug: "pm-table(:current-page='page' :per-page='10' :busy='loading' @sort-change='onSort' @page-change='onPage' @filter-change='onFilter')",
        description: 'Handle all server interactions on change handlers.',
      },
    ],
    restrictions: [
      { type: 'runtime', message: 'Must use with backend paging/sorting API endpoint that honors query params.' },
      { type: 'runtime', message: 'Server must return `items` and `totalRows` consistently.' },
      { type: 'security', message: 'Validate all query params before forwarding to backend data source.' },
    ],
    enabled: true,
    version: '0.4.2',
  },
  {
    name: 'DropzoneUploader',
    module: 'vue-dropzone',
    tag: 'DropzoneUploader',
    pack: 'files',
    props: [
      { name: 'uploadUrl', type: 'string', required: true, description: 'Target upload endpoint.' },
      { name: 'maxFileSize', type: 'number', required: false, defaultValue: '10485760', description: 'Max bytes per file.' },
      { name: 'acceptedFiles', type: 'string', required: false, defaultValue: '*/*', description: 'Accepted MIME types.' },
      { name: 'multiple', type: 'boolean', required: false, defaultValue: 'true', description: 'Allow multiple uploads.' },
    ],
    slots: [
      { name: 'thumb', description: 'Custom preview template.' },
      { name: 'status', description: 'Upload status and error messages.' },
    ],
    events: [
      { name: 'upload-success', payload: '{ id: string, file: object }', description: 'File uploaded successfully.' },
      { name: 'upload-error', payload: '{ id: string, error: string }', description: 'Upload failed.' },
      { name: 'update:files', payload: 'Array<object>', description: 'Selected/uploaded list changed.' },
    ],
    examples: [
      {
        label: 'Avatar uploader',
        pug: "DropzoneUploader(upload-url='/api/files' :max-file-size='5242880' accepted-files='image/*' :multiple='false')",
        description: 'Single image upload with MIME filtering.',
      },
    ],
    restrictions: [
      { type: 'runtime', message: 'Uploads must be validated server-side before persistence.' },
      { type: 'security', message: 'Only allow known-safe MIME types and sanitize file names.' },
    ],
    enabled: true,
    version: '0.9.8',
  },
];

File: apps/api/internal/registry/seed.go
package registry

const CatalogVersion = "2026-04-25"

var initialComponentInventory = []ComponentInventoryItem{
	{
		Name:   "DateRangePicker",
		Module: "advanced-inputs",
		Tag:    "DateRangePicker",
		Pack:   "advanced-inputs",
		Props: []ComponentPropMetadata{
			{Name: "modelValue", Type: "string | [Date, Date] | null", Required: false, Description: strPtr("Current selected date range.")},
			{Name: "placeholder", Type: "string", Required: false, Default: strPtr("Start date - End date"), Description: strPtr("Input placeholder text.")},
			{Name: "clearable", Type: "boolean", Required: false, Default: strPtr("true"), Description: strPtr("Show clear button.")},
			{Name: "format", Type: "string", Required: false, Default: strPtr("YYYY-MM-DD"), Description: strPtr("Date format passed to formatter.")},
		},
		Slots: []ComponentSlotMetadata{
			{Name: "header", Description: strPtr("Replace the calendar header area." )},
			{Name: "footer", Description: strPtr("Append custom action buttons.")},
		},
		Events: []ComponentEventMetadata{
			{Name: "update:modelValue", Payload: strPtr("{ start?: string; end?: string }"), Description: strPtr("Emits when range changes.")},
			{Name: "apply", Payload: strPtr("{ start: string; end: string }"), Description: strPtr("Emits when user applies a range.")},
		},
		Examples: []ComponentExample{
			{Label: "Report period chooser", Pug: "DateRangePicker(v-model='reportRange' :clearable='true' format='YYYY-MM-DD')", Description: strPtr("Basic form usage in form groups.")},
		},
		Restrictions: []ComponentRestriction{
			{Type: RestrictionRuntime, Message: "Should not be used inside SSR-only rendering paths."},
			{Type: RestrictionUX, Message: "Pair with clear reset controls when used in filter forms."},
		},
		Enabled: true,
		Version: strPtr("1.2.0"),
	},
	{
		Name:   "AsyncMultiSelect",
		Module: "advanced-inputs",
		Tag:    "AsyncMultiSelect",
		Pack:   "advanced-inputs",
		Props: []ComponentPropMetadata{
			{Name: "modelValue", Type: "string[]", Required: false, Description: strPtr("Selected IDs.")},
			{Name: "search", Type: "(query: string) => Promise<Option[]>", Required: true, Description: strPtr("Async remote query function.")},
			{Name: "debounceMs", Type: "number", Required: false, Default: strPtr("300"), Description: strPtr("Delay between remote requests.")},
			{Name: "labelKey", Type: "string", Required: false, Default: strPtr("label"), Description: strPtr("Option label field.")},
			{Name: "valueKey", Type: "string", Required: false, Default: strPtr("value"), Description: strPtr("Option value field.")},
		},
		Slots: []ComponentSlotMetadata{
			{Name: "option", Description: strPtr("Custom rendering for each suggestion item.")},
			{Name: "selected", Description: strPtr("Custom rendering for selected tokens.")},
		},
		Events: []ComponentEventMetadata{
			{Name: "search", Payload: strPtr("string"), Description: strPtr("Fires whenever query changes.")},
			{Name: "update:modelValue", Payload: strPtr("string[]"), Description: strPtr("Selection changes.")},
		},
		Examples: []ComponentExample{
			{Label: "Team member picker", Pug: "AsyncMultiSelect(v-model='selectedUsers' :search='fetchUsers' label-key='fullName' value-key='id')", Description: strPtr("Load options as user types.")},
		},
		Restrictions: []ComponentRestriction{
			{Type: RestrictionRuntime, Message: "Requires a function prop that resolves within 15s."},
			{Type: RestrictionSecurity, Message: "Search input is sanitized through backend API proxy."},
		},
		Enabled: true,
		Version: strPtr("1.2.0"),
	},
	{
		Name:   "InputMask",
		Module: "advanced-inputs",
		Tag:    "InputMask",
		Pack:   "advanced-inputs",
		Props: []ComponentPropMetadata{
			{Name: "modelValue", Type: "string", Required: false, Description: strPtr("Masked string value.")},
			{Name: "mask", Type: "string", Required: true, Description: strPtr("Mask pattern, e.g. (999) 999-9999")},
			{Name: "placeholder", Type: "string", Required: false, Default: strPtr("Enter value"), Description: strPtr("Placeholder while empty.")},
			{Name: "unmask", Type: "boolean", Required: false, Default: strPtr("false"), Description: strPtr("Return unmasked value on input.")},
		},
		Slots: []ComponentSlotMetadata{
			{Name: "addon", Description: strPtr("Custom input addon slot.")},
		},
		Events: []ComponentEventMetadata{
			{Name: "complete", Payload: strPtr("string"), Description: strPtr("All mask characters completed.")},
			{Name: "update:modelValue", Payload: strPtr("string"), Description: strPtr("Output value updates.")},
		},
		Examples: []ComponentExample{
			{Label: "Phone field", Pug: "InputMask(v-model='phone' mask='(999) 999-9999' placeholder='(201) 555-0123')", Description: strPtr("Use for strict format fields.")},
		},
		Restrictions: []ComponentRestriction{
			{Type: RestrictionUX, Message: "Use with short fields and helper text for expected format."},
			{Type: RestrictionSecurity, Message: "Do not bind secrets or tokens to this component."},
		},
		Enabled: true,
		Version: strPtr("1.2.0"),
	},
	{
		Name:   "pm-table",
		Module: "packages/component-registry",
		Tag:    "pm-table",
		Pack:   "files",
		Props: []ComponentPropMetadata{
			{Name: "modelValue", Type: "Array<Record<string, any>>", Required: false, Description: strPtr("Rows being displayed.")},
			{Name: "totalRows", Type: "number", Required: false, Description: strPtr("Total matching rows from server.")},
			{Name: "sortBy", Type: "string[]", Required: false, Description: strPtr("Sort columns order.")},
			{Name: "sortDesc", Type: "boolean[]", Required: false, Description: strPtr("Sort directions.")},
			{Name: "currentPage", Type: "number", Required: false, Default: strPtr("1"), Description: strPtr("Current page index.")},
			{Name: "perPage", Type: "number", Required: false, Default: strPtr("10"), Description: strPtr("Page size.")},
			{Name: "filter", Type: "Record<string, any>", Required: false, Description: strPtr("Server-side filter payload.")},
			{Name: "busy", Type: "boolean", Required: false, Default: strPtr("false"), Description: strPtr("Loading state.")},
		},
		Slots: []ComponentSlotMetadata{
			{Name: "cell", Description: strPtr("Custom cell rendering.")},
			{Name: "empty", Description: strPtr("Shown when dataset is empty.")},
		},
		Events: []ComponentEventMetadata{
			{Name: "sort-change", Payload: strPtr("{ sortBy: string[], sortDesc: boolean[] }"), Description: strPtr("User changed sorting criteria.")},
			{Name: "filter-change", Payload: strPtr("Record<string, any>"), Description: strPtr("User changed filter values.")},
			{Name: "page-change", Payload: strPtr("number"), Description: strPtr("Page number changed.")},
			{Name: "update:items", Payload: strPtr("Array<Record<string, any>>"), Description: strPtr("Updated data slice received from server.")},
		},
		Examples: []ComponentExample{
			{Label: "Server paginated list", Pug: "pm-table(:current-page='page' :per-page='10' :busy='loading' @sort-change='onSort' @page-change='onPage' @filter-change='onFilter')", Description: strPtr("Handle all server interactions on change handlers.")},
		},
		Restrictions: []ComponentRestriction{
			{Type: RestrictionRuntime, Message: "Must use with backend paging/sorting API endpoint that honors query params."},
			{Type: RestrictionRuntime, Message: "Server must return `items` and `totalRows` consistently."},
			{Type: RestrictionSecurity, Message: "Validate all query params before forwarding to backend data source."},
		},
		Enabled: true,
		Version: strPtr("0.4.2"),
	},
	{
		Name:   "DropzoneUploader",
		Module: "vue-dropzone",
		Tag:    "DropzoneUploader",
		Pack:   "files",
		Props: []ComponentPropMetadata{
			{Name: "uploadUrl", Type: "string", Required: true, Description: strPtr("Target upload endpoint.")},
			{Name: "maxFileSize", Type: "number", Required: false, Default: strPtr("10485760"), Description: strPtr("Max bytes per file.")},
			{Name: "acceptedFiles", Type: "string", Required: false, Default: strPtr("*/*"), Description: strPtr("Accepted MIME types.")},
			{Name: "multiple", Type: "boolean", Required: false, Default: strPtr("true"), Description: strPtr("Allow multiple uploads.")},
		},
		Slots: []ComponentSlotMetadata{
			{Name: "thumb", Description: strPtr("Custom preview template.")},
			{Name: "status", Description: strPtr("Upload status and error messages.")},
		},
		Events: []ComponentEventMetadata{
			{Name: "upload-success", Payload: strPtr("{ id: string, file: object }"), Description: strPtr("File uploaded successfully.")},
			{Name: "upload-error", Payload: strPtr("{ id: string, error: string }"), Description: strPtr("Upload failed.")},
			{Name: "update:files", Payload: strPtr("Array<object>"), Description: strPtr("Selected/uploaded list changed.")},
		},
		Examples: []ComponentExample{
			{Label: "Avatar uploader", Pug: "DropzoneUploader(upload-url='/api/files' :max-file-size='5242880' accepted-files='image/*' :multiple='false')", Description: strPtr("Single image upload with MIME filtering.")},
		},
		Restrictions: []ComponentRestriction{
			{Type: RestrictionRuntime, Message: "Uploads must be validated server-side before persistence."},
			{Type: RestrictionSecurity, Message: "Only allow known-safe MIME types and sanitize file names."},
		},
		Enabled: true,
		Version: strPtr("0.9.8"),
	},
	{
		Name:   "VueChart",
		Module: "vue-chartjs",
		Tag:    "VueChart",
		Pack:   "charts",
		Props: []ComponentPropMetadata{
			{Name: "chartType", Type: "'line' | 'bar' | 'pie' | 'doughnut' | 'radar' | 'polarArea' | 'bubble' | 'scatter'", Required: false, Default: strPtr("line"), Description: strPtr("Chart.js chart type to render.")},
			{Name: "type", Type: "string", Required: false, Description: strPtr("Alias for chartType.")},
			{Name: "chartData", Type: "Record<string, any> | null", Required: true, Description: strPtr("Data object for vue-chartjs/chart.js")},
			{Name: "chartOptions", Type: "Record<string, any>", Required: false, Description: strPtr("Options object for chart.js.")},
			{Name: "options", Type: "Record<string, any>", Required: false, Description: strPtr("Alias for chartOptions.")},
			{Name: "width", Type: "number | string", Required: false, Description: strPtr("Optional rendered chart width.")},
			{Name: "height", Type: "number | string", Required: false, Description: strPtr("Optional rendered chart height.")},
		},
		Slots:  []ComponentSlotMetadata{},
		Events: []ComponentEventMetadata{},
		Examples: []ComponentExample{
			{
				Label:       "Sales chart",
				Pug:         "VueChart(chart-type='line' :chart-data='salesData' :chart-options='salesOptions')",
				Description: strPtr("Render a line chart bound to data variables."),
			},
		},
		Restrictions: []ComponentRestriction{
			{Type: RestrictionRuntime, Message: "chart-data must be compatible with vue-chartjs datasets format for the selected chart-type."},
			{Type: RestrictionSecurity, Message: "Avoid untrusted config objects in chartOptions and keep values JSON serializable."},
		},
		Enabled: true,
		Version: strPtr("1.0.0"),
	},
}

func strPtr(value string) *string {
	v := value
	return &v
}

